src/replay/scheduler.zig

```diff
@@ -0,0 +1,294 @@
+const std = @import("std");
+const sig = @import("../sig.zig");
+const replay = @import("lib.zig");
+
+const Allocator = std.mem.Allocator;
+
+const Channel = sig.sync.Channel;
+const HomogeneousThreadPool = sig.utils.thread.HomogeneousThreadPool;
+const ThreadPool = sig.sync.ThreadPool;
+
+const TransactionError = sig.ledger.transaction_status.TransactionError;
+
+const AccountLocks = replay.account_locks.AccountLocks;
+const ConfirmSlotStatus = replay.confirm_slot.ConfirmSlotStatus;
+const ResolvedTransaction = replay.resolve_lookup.ResolvedTransaction;
+const ResolvedBatch = replay.resolve_lookup.ResolvedBatch;
+
+const ScopedLogger = sig.trace.ScopedLogger("replay-batcher");
+
+const assert = std.debug.assert;
+
+/// Processes a batch of transactions by verifying their signatures and
+/// executing them with the SVM.
+pub fn processBatch(
+    transactions: []const ResolvedTransaction,
+    exit: *std.atomic.Value(bool),
+) ?TransactionError {
+    for (transactions) |transaction| {
+        if (exit.load(.monotonic)) {
+            return null;
+        }
+        _ = transaction.transaction.verifyAndHashMessage() catch return .SignatureFailure;
+        // TODO: call svm
+    }
+    // TODO: commit results
+
+    return null;
+}
+
+/// Batches transactions and executes them on a thread pool.
+///
+/// Internally, transaction results are communicated on a channel instead of
+/// collecting the return values through HomogeneousThreadPool to allow faster
+/// identification of failed transactions when batches are completed out of
+/// order.
+///
+/// Transaction errors are not returned by the task. If a task returns an error,
+/// that indicates fatal/unexpected errors like OOM, and the error will
+/// propagate up when polling.
+///
+/// This should only be used in a single thread at a time.
+pub const TransactionScheduler = struct {
+    allocator: Allocator,
+    batches: std.ArrayListUnmanaged(ResolvedBatch),
+    thread_pool: HomogeneousThreadPool(ProcessBatchTask),
+    results: Channel(?TransactionError),
+    locks: AccountLocks,
+    /// The number of batches that have been scheduled with thread_pool.trySchedule.
+    batches_started: usize,
+    /// The number of batches that a result has been received over the channel for.
+    batches_finished: usize,
+    /// triggered as soon as a single transaction fails
+    exit: std.atomic.Value(bool),
+    /// if non-null, a failure was already recorded and will be returned for every poll
+    failure: ?replay.confirm_slot.ConfirmSlotError,
+
+    pub fn initCapacity(
+        allocator: Allocator,
+        batch_capacity: usize,
+        thread_pool: *ThreadPool,
+    ) !TransactionScheduler {
+        var batches = try std.ArrayListUnmanaged(ResolvedBatch)
+            .initCapacity(allocator, batch_capacity);
+        errdefer batches.deinit(allocator);
+
+        const pool = try HomogeneousThreadPool(ProcessBatchTask)
+            .initBorrowed(allocator, thread_pool, null);
```

Comment:
```suggestion
        const pool: HomogeneousThreadPool(ProcessBatchTask) =
            try .initBorrowed(allocator, thread_pool, null);
```

Reply:
I prefer the current style and don't plan to change this.